### PR TITLE
Tests for Super1155, Staker, MintShop1155 with base contract changes

### PR DESCRIPTION
**Pull request by qzirak**
=====Tests Written for Contracts
- Staker.sol
- Super1155.sol
- MintShop1155.sol
- PermitControl.sol
- Sweepable.sol

=====Mockfiles included for Tests
- MockERC20.sol
- MockERC1155Receivers.sol

=====Base contracts modified
Staker.sol
- setEmissions()

MintShop1155.sol
- getWhitelistStatus()
- getPurchaseCounts()

Sweepable.sol
- sweep()

=====Some improvements to consider
- Adding a revert string to the modifier (Super1155.sol)

=====Statements that could not be covered
Super1155.sol
- _doSafeBatchTransferAcceptanceCheck() => catch(Error): Unable to produce the Error for testing
- _doSafeTransferAcceptanceCheck() => catch(Error): Unable to produce the Error for testing
- line 711 => itemGroups[_groupId].itemType == ItemType.Fungible: Enum type check is done when creating the group. Invalid Enum value assignment throws an exception so the else part is never run.

MintShop1155.sol
- line 1043 => else part of this require => "MintShop1155: payment receiver transfer failed": Unable to produce the Error for testing
- line 1047 => sellingPair.assetType == AssetType.Token: Enum type check is done at the time of assigning value to it. Invalid Enum value assignment throws an exception so the else part's revert("MintShop1155: unrecognized asset type") is never run.

PermitControl.sol
- version(): Overridden

Sweepable.sol
- version(): Overridden

=====HardhatCoverage
- Make sure to do "npx hardhat clean" before running "npx hardhat coverage"

### CODE REVIEW

File: contracts/MintShop1155.sol [hunk 1]
```diff
@@ -678,14 +679,14 @@ contract MintShop1155 is Sweepable, ReentrancyGuard {
...
     for (uint256 i = 0; i < _ids.length; i++) {
+      whitelistStatus[i] = new bool[](_addresses.length);
       uint256 id = _ids[i];
       uint256 whitelistVersion = whitelists[id].currentWhitelistVersion;
       for (uint256 j = 0; j < _addresses.length; j++) {
-        bytes32 addressKey = keccak256(abi.encode(whitelistVersion,
-          _addresses[j]));
-        whitelistStatus[j][i] = whitelists[id].addresses[addressKey];
+        bytes32 addressKey = keccak256(abi.encode(whitelistVersion, _addresses[j]));
+        whitelistStatus[i][j] = whitelists[id].addresses[addressKey];
```
Comment (TimTinkers) [inline contracts/MintShop1155.sol:689]: Good catch with the mapping order being reversed. Our front-end needs are such that it's preferable to receive data structured where we can index with the address as the first field, but this reversal made it hard to correlate the output with the input. I'd like to keep the mapping its current address-first id-second ordering, but rearrange the ordering of the `_addresses` and `_ids` arrays to make that output make more sense.

File: contracts/MintShop1155.sol [hunk 2]
```diff
@@ -752,12 +753,13 @@ contract MintShop1155 is Sweepable, ReentrancyGuard {
...
     address[] calldata _purchasers) external view returns (uint256[][] memory) {
-    uint256[][] memory purchaseCounts;
+    uint256[][] memory purchaseCounts = new uint256[][](_ids.length);
     for (uint256 i = 0; i < _ids.length; i++) {
+      purchaseCounts[i] = new uint256[](_purchasers.length);
       uint256 id = _ids[i];
       for (uint256 j = 0; j < _purchasers.length; j++) {
         address purchaser = _purchasers[j];
-        purchaseCounts[j][i] = pools[id].purchaseCounts[purchaser];
+        purchaseCounts[i][j] = pools[id].purchaseCounts[purchaser];
```
Comment (TimTinkers) [inline contracts/MintShop1155.sol:762]: Ditto here; I'd prefer we keep the same output order with input parameters as `_purchasers` then `_ids`.

File: contracts/Staker.sol [hunk 3]
```diff
@@ -183,7 +184,7 @@ contract Staker is Ownable, ReentrancyGuard {
     Set new emission details to the Staker or overwrite existing ones.
     This operation requires that emission schedule alteration is not locked.
 
-    @param _tokenSchedule An array of EmissionPoints defining the token schedule.
+    @param _tokenSchedule An array of EmissionTokens defining the token schedule.
```
Comment (TimTinkers) [inline contracts/Staker.sol:187]: There's not actually a struct called `EmissionTokens`; the "Point" in `EmissionPoints` is supposed to refer to a point of change on some emissions curve.

File: contracts/Super1155.sol [hunk 4]
```diff
@@ -320,7 +320,7 @@ contract Super1155 is PermitControl, ERC165, IERC1155, IERC1155MetadataURI {
     } else if (hasRightUntil(_msgSender(), bytes32(_id), _right)
       > block.timestamp) {
       _;
-    }
+    }
```
Comment (TimTinkers) [inline contracts/Super1155.sol:323]: Added a single space? Let's cut this change out of the PR to keep the commit history as simple as possible.

**Review by TimTinkers — CHANGES_REQUESTED** [review]

File: contracts/Super1155.sol (hunk 4 above)
Comment (qzirak) [reply contracts/Super1155.sol:323]: Alright. This was to indicate the place for the revert statement. But I'll remove this

**Pull request closed without merging**